sudoers_helper.c

```diff
@@ -0,0 +1,256 @@
+#include <stdarg.h>
+#include <stdio.h>
+#include <string.h>
+#include <syslog.h>
+#include <stdlib.h>
+#include <stdbool.h>
+
+#include "password.h"
+#include "sudoers_helper.h"
+#include "trace.h"
+
+/* Macros for parse user input */
+#define USER_COMMAND_TOKEN_WHITESPACE              " \t\n\r\f"
+#define USER_COMMAND_TOKEN_SETTING_SPLITTER        " =\t"
+#define USER_COMMAND_TOKEN_EQUAL                   "="
+#define USER_COMMAND_TOKEN_COMMA                   ","
+
+#define BUFFER_CRLF              "\n\r"
+
+/* The command alias prefix */
+const char* COMMAND_ALIAS = "Cmnd_Alias";
+
+/* The password setting */
+const char* PASSWD_CMDS = "PASSWD_CMDS";
+
+/* Append string to buffer. */
+char* append_string_to_buffer(char* buffer, char* str)
+{
+    int str_len = strlen(str);
+    int buffer_len = 0;
+    if (buffer != NULL) {
+        buffer_len = strlen(buffer);
+    }
+
+    buffer = realloc(buffer, buffer_len + str_len + 1);
+
+    memcpy(buffer + buffer_len, str, str_len);
+    buffer[buffer_len + str_len] = 0;
+
+    return buffer;
+}
+
+/*
+    Load PASSWD_CMDS from sudoers.
+    For more information please check:
+        https://www.sudo.ws/man/1.8.17/sudoers.man.html#Other_special_characters_and_reserved_words
+*/
+char* load_passwd_cmds(const char *setting_path)
+{
+    FILE *setting_file= fopen(setting_path, "r");
+    if(setting_file == NULL) {
+        trace("Can't open setting file: %s\n", setting_path);
+        return NULL;
+    }
+
+    int continue_load_password = NO_NEXT_SETTING_LINE;
+    char *passwd_cmds = NULL;
+    char line_buffer[MAX_LINE_SIZE+1];
+    while (fgets(line_buffer, sizeof line_buffer, setting_file)) {
+        /* Remove \r\n from buffer */
+        line_buffer[strcspn(line_buffer, BUFFER_CRLF)] = 0;
```

Comment:
>BUFFER_CRLF [](http://example.com/codeflow?start=41&length=11)

Did you test on Linux? Suggest open file as `rt` mode, so only scan `\n`.

Reply:
Fixed.

Reply:
Sorry I double checked, `t` is the default mode and could be implicit.